exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    /// #       public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ExecutionError {
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    code: u8,
+    /// Optional error description.
+    description: Option<String>,
+}
+
+impl ExecutionError {
+    /// Constructs a new `ExecutionError` instance with the given error code.
+    pub fn new(code: u8) -> Self {
+        Self {
+            code,
+            description: None,
+        }
+    }
+
+    /// Constructs a new `ExecutionError` instance with the given error code and description.
+    pub fn with_description(code: u8, description: String) -> Self {
+        Self {
+            code,
+            description: Some(description),
+        }
+    }
+}
+
+/// Type of the transaction error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionErrorType {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+}
+
+/// Extended by the framework result of unsuccessful transaction execution.
+///
+/// # Notes:
+///
+/// - Content of `description`' field is excluded from hash calculation (see `StorageValue`
+///   implementation for the details).
+/// - `TransactionErrorType::Panic` is set by the framework if panic is raised during transaction
+///   execution.
+/// - `TransactionError` implements `Display` which can be used for obtaining a simple error
+///   description.
+///
+/// # Examples
+///
+/// ```
+/// # use exonum::blockchain::{ExecutionError, TransactionError};
+///
+/// # let transaction_error: TransactionError = ExecutionError::new(1).into();
+/// // Prints user friendly error description.
+/// println!("Transaction error: {}", transaction_error);
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TransactionError {
+    /// Error type, see `TransactionErrorType` for the details.
+    error_type: TransactionErrorType,
```

Comment:
IMO, naming this field `type` could be better; it avoids stuttering.

Reply:
`type` is a keyword and I really don't like things like `type_`.

Reply:
Jeez, I sure didn't think that through... OTOH, imo `error_type` is close in its semantic to a *code* (see a comment about `TransactionErrorType` naming), so maybe that could be used?

Reply:
Personally, I like `error_type` more, lets see other comments if there are any.

Reply:
I don't think naming of a single private field can really be a problem.

Reply:
There is also a public getter for the field and the type of the field, but I agree that it's not that important.